### PR TITLE
Get user location

### PR DESCRIPTION
**Pull request by sbotichelli**
1) requested user location and updated map position on it's basic. 
TODO in this pull request: get the code out of MapView.swift

### CODE REVIEW

**Comment by sbotichelli** [comment]
User location service interactions are moved to separate a class.

**Comment by sbotichelli** [comment]
TODO: fix the problem that map region doesn't change when user accepts tracking his GPS position.

File: BusTracker/Coordinators/AppCoordinator.swift [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        window.makeKeyAndVisible()
+    }
+}
+
+private extension AppCoordinator {
+    func updateCurrentCoordinator() {
+        let navigation = UINavigationController()
+        window.rootViewController = navigation
+        
+        let running = RunningCoordinator(navigation: navigation)
```
Comment (SevInf) [inline BusTracker/Coordinators/AppCoordinator.swift]: You can assign this directly to `coorditor`, without intermediate variable

File: BusTracker/Location/UserLocation.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    let locationManager = CLLocationManager()
+    var delegate: UserLocationDelegate?
+    let defaultRegionSpanInMeteres = CLLocationDistance(1000)
+    
+    override init() {
+        super.init()
+    }
+    
+    func checkLocationAuthorisationStatus() {
+        locationManager.delegate = self
```
Comment (SevInf) [inline BusTracker/Location/UserLocation.swift]: Can be done in `init` — location managers delegate will always be `self`, we don't need to reset it every time `checkLocationAuthorisationStatus` is called.

File: BusTracker/Location/UserLocation.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+//
+//  Created by botichelli on 9/24/19.
+//  Copyright © 2019 Oleksandra Sildushkina. All rights reserved.
+//
+
+import MapKit
+
+protocol UserLocationDelegate {
+    func setMapRegion(region: MKCoordinateRegion)
+    func showTrackingButton()
```
Comment (SevInf) [inline BusTracker/Location/UserLocation.swift]: I don't think this naming helps with separation of concerns. Service classes are not supposed to know anything about the UI. Naming this method `showTrackingButton` makes user location class know too much about its usage. From the code below, it looks like what it is actually reporting is the fact that user authorized location access for the app. Would be nice to change the method name to convey that instead. This way, if you decide to do something else in reaction to authorization or reuse `UserLocation` in some different place, you'll need to change only controllers/view, `UserLocation` class will stay as it is.

File: BusTracker/Networking/APIResponceParser.swift [hunk 4]
```diff
@@ -9,53 +9,126 @@
...
-                case "Błędna metoda lub parametry wywołania":
-                    print("No parameter: resource_id")
-                default:
-                    print("What's going on???")
-                }
-            } else {
-                
-                let vehiclePositions = responce["result"] as? [Any] ?? []
+        do {
+            if let responce = try JSONSerialization.jsonObject(with: APIResponce.data, options: []) as? [String : Any] {
```
Comment (SevInf) [inline BusTracker/Networking/APIResponceParser.swift]: `response`

File: BusTracker/Networking/APIResponceParser.swift [hunk 5]
```diff
@@ -9,53 +9,126 @@
...
+                wereParsingErrors = true
+            }
+        }
+        if parcedVehiclePositions.isEmpty && wereParsingErrors {
+            throw APIResponceParserError.wrongJSONWithParameters(list: erroringParameters)
+        }
         return parcedVehiclePositions
     }
+    
+    func parceVehicle(data: [String : Any], type: VehicleType) throws -> TransportPosition {
```
Comment (SevInf) [inline BusTracker/Networking/APIResponceParser.swift]: `parse`

File: BusTracker/Servis/RepeatingTimer.swift [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            return
+        }
+        state = .suspended
+        timer.suspend()
+    }
+    
+    deinit {
+        timer.setEventHandler {}
+        timer.cancel()
+        resume()
```
Comment (SevInf) [inline BusTracker/Servis/RepeatingTimer.swift]: That looks weird to me. Why do we want to resume the timer in destructor?

File: BusTracker/ViewControllers/MapVC/View/MapView.swift [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        userLocation.delegate = self
+        setupView()
+        registerAnnotations()
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("Init from NIB is not emplemented.")
```
Comment (SevInf) [inline BusTracker/ViewControllers/MapVC/View/MapView.swift]: `implemented`

File: BusTracker/ViewControllers/MapVC/View/MapView.swift [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import Foundation
+import UIKit
+import SnapKit
+import MapKit
+
+class MapView: UIView {
+    
+    let mapView = MKMapView()
+    private var userTrackingButton: MKUserTrackingButton!
+    let defaultRegionSpanInMeteres = CLLocationDistance(1000)
```
Comment (SevInf) [inline BusTracker/ViewControllers/MapVC/View/MapView.swift]: This constant is duplicated between several classes, would be nice to put it into a single place.

File: BusTracker/ViewControllers/MapVC/View/MapView.swift [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        userLocation.delegate = self
+        setupView()
+        registerAnnotations()
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("Init from NIB is not emplemented.")
+    }
+    
+    func setupView() {
```
Comment (SevInf) [inline BusTracker/ViewControllers/MapVC/View/MapView.swift]: Is there a reason for having this method? IMHO, either `setupView` or `setupMapView` could be removed and remaining method could do all the work.

File: BusTracker/ViewControllers/MapVC/View/MapView.swift [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        userTrackingButton.isHidden = true
+        mapView.addSubview(userTrackingButton)
+        userTrackingButton.snp.makeConstraints { (make) in
+            make.right.equalToSuperview().offset(-20)
+            make.bottom.equalToSuperview().offset(-20)
+        }
+    }
+}
+
+extension MapView: UserLocationDelegate {
```
Comment (SevInf) [inline BusTracker/ViewControllers/MapVC/View/MapView.swift]: Generally, I'd say controller is supposed to be a mediator between service classes and the view. So, I'd redo it in the following way:

1. Controller is the delegate for UserLocation.
2. View has methods for updating the region and showing button.
3. Controller calls appropriate view methods  in response to events from `UserLocation`.
4. View and `UserLocation` don't know about each other.

Would play nicely with `UserLocationDelegate` renames I suggested earlier.

Generally speaking, separation should work somewhat this way:
1. Service classes - tells controller when something important happens.
2. Views - show pretty pictures.
3. Controllers - tells view, which pretty picture to show when service tells it that something important happened.

File: BusTracker/Networking/APIResponceParser.swift [hunk 11]
```diff
@@ -9,53 +9,126 @@
...
-                    let parcedVehicle = TransportPosition(line: vehicleData["Lines"] as! String,
-                                                          latitude: vehicleData["Lat"] as! Double,
-                                                          longitude: vehicleData["Lon"] as! Double,
-                                                          brigade: vehicleData["Brigade"] as! String,
-                                                          updateTime: positionUpdateDatetime,
-                                                          vehicleType: vehicleType)
-                    
-                    parcedVehiclePositions.append(parcedVehicle)
+                // Interpret error message from API
+                if let message = responce["result"] as? String {
```
Comment (SevInf) [inline BusTracker/Networking/APIResponceParser.swift]: I wonder if some third party library would make parsing less verbose.

File: BusTracker/Location/UserLocation.swift [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+            locationManager.desiredAccuracy = kCLLocationAccuracyNearestTenMeters
+            locationManager.startUpdatingLocation()
+        } else {
+            locationManager.requestWhenInUseAuthorization()
+        }
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
+        guard let locValue: CLLocationCoordinate2D = manager.location?.coordinate else {return}
+        let region = MKCoordinateRegion(center: locValue, latitudinalMeters: defaultRegionSpanInMeteres, longitudinalMeters: defaultRegionSpanInMeteres)
```
Comment (SevInf) [inline BusTracker/Location/UserLocation.swift]: I'd consider moving this calculation either to controller or to view and report a single coordinate from `UserLocation`.

File: BusTracker/Location/UserLocation.swift [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        delegate?.setMapRegion(region: region)
+        locationManager.stopUpdatingLocation()
```
Comment (SevInf) [inline BusTracker/Location/UserLocation.swift]: Do we really want to stop location tracking immediately after we got first position?

File: BusTracker/Servis/RepeatingTimer.swift (hunk 6 above)
Comment (sbotichelli) [reply BusTracker/Servis/RepeatingTimer.swift]: Copied code 😅 Changed to 
`state = .suspended`

File: BusTracker/ViewControllers/MapVC/View/MapView.swift (hunk 9 above)
Comment (sbotichelli) [reply BusTracker/ViewControllers/MapVC/View/MapView.swift]: Currently there's only a map view on screen. I'm planning to add some navbar buttons. So this method will get extended for sure.

File: BusTracker/Networking/APIResponceParser.swift (hunk 11 above)
Comment (sbotichelli) [reply BusTracker/Networking/APIResponceParser.swift]: It's an issue for the next pull request. Saving for later.

**Pull request merged**